### PR TITLE
feat: implement custom docs layout

### PR DESCRIPTION
**Pull request by zchsh**
## 🔗 Relevant links

- [Preview link][preview] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Implements hero sections for all docs landing pages.

## 🤷 Why

To match design spec.

## 🛠️ How

- Carries forward approach from #733
- Adds optional `description` property on `rootDocsPath` entries, to allow `subheading` for `LandingHero` to be authorable (falls back to `frontmatter.description` from the MDX file)
- Tweaks spacing approach to match design spec

### Context

CleanShot 2022-08-08 at 14 14 35@2x

#### Typical example

CleanShot 2022-08-23 at 17 39 10@2x

#### `/docs` custom landing page, using same component

CleanShot 2022-08-23 at 17 39 28@2x

#### Example where page has no `h1`, so no `LandingHero`

CleanShot 2022-08-23 at 17 39 38@2x

#### Vault pages, which are in public beta

CleanShot 2022-08-23 at 17 40 28@2x

CleanShot 2022-08-23 at 17 40 16@2x

#### Waypoint pages, which are in public beta

CleanShot 2022-08-23 at 17 39 57@2x

CleanShot 2022-08-24 at 09 45 34@2x

CleanShot 2022-08-24 at 09 45 47@2x

## 🧪 Testing

> **Note**: docs pages marked with ⭐️ are currently in public beta, and therefore likely merit extra attention.

- [ ] Visit docs landing pages for all other products, and ensure their appearance is as expected
  - Boundary
    - [/boundary/api-docs][/boundary/api-docs]
        > **Note**: does not use the standard docs template, listed for completeness
    - [/boundary/docs][/boundary/docs]
  - Consul
    - [/consul/api-docs][/consul/api-docs]
    - [/consul/commands][/consul/commands]
    - [/consul/docs][/consul/docs]
  - HCP
    - [/hcp/docs][/hcp/docs]
  - Nomad
    - [/nomad/api-docs][/nomad/api-docs]
    - [/nomad/docs][/nomad/docs]
    - [/nomad/intro][/nomad/intro]
    - [/nomad/plugins][/nomad/plugins]
    - [/nomad/tools][/nomad/tools]
  - Packer
    - [/packer/docs][/packer/docs]
    - [/packer/guides][/packer/guides]
    - [/packer/intro][/packer/intro]
    - [/packer/plugins][/packer/plugins]
  - Terraform
    - [/terraform/cdktf][/terraform/cdktf]
    - [/terraform/cli][/terraform/cli]
    - [/terraform/cloud-docs][/terraform/cloud-docs]
    - [/terraform/cloud-docs/agents][/terraform/cloud-docs/agents]
    - [/terraform/docs][/terraform/docs]
    - [/terraform/enterprise][/terraform/enterprise]
    - [/terraform/internals][/terraform/internals]
    - [/terraform/intro][/terraform/intro]
    - [/terraform/language][/terraform/language]
    - [/terraform/plugin][/terraform/plugin]
    - [/terraform/plugin/framework][/terraform/plugin/framework]
    - [/terraform/plugin/log][/terraform/plugin/log]
      > **Note**: page does not contain an `<h1>`, so we have nothing to render in the hero, and therefore skip rendering it. This could be resolved by changing the `h2` on the page to an `h1`. We likely want to make this content fix regardless.
    - [/terraform/plugin/mux][/terraform/plugin/mux]
    - [/terraform/plugin/sdkv2][/terraform/plugin/sdkv2]
    - [/terraform/registry][/terraform/registry]
  - Vagrant
    - [/vagrant/docs][/vagrant/docs]
    - [/vagrant/intro][/vagrant/intro]
    - [/vagrant/vagrant-cloud][/vagrant/vagrant-cloud]
  - ⭐️ Vault
    - ⭐️ [/vault/docs][/vault/docs]
    - ⭐️ [/vault/api-docs][/vault/api-docs]
  - ⭐️ Waypoint
    - ⭐️ [/waypoint/commands][/waypoint/commands]
       > **Note**: no frontmatter `description` is defined, so we should the `heading` only. This could be fixed in this PR by defining a `description` property on the `"path": "commands"` `rootDocsPath` in `waypoint.json`.
    - ⭐️ [/waypoint/docs][/waypoint/docs]
    - ⭐️ [/waypoint/plugins][/waypoint/plugins]
      > **Note**: no frontmatter `description` is defined, so we should the `heading` only. This could be fixed in this PR by defining a `description` property on the `"path": "plugins"` `rootDocsPath` in `waypoint.json`.

## 💭 Anything else?

⭐ Note that this change affects the live-in-beta `Waypoint` and `Vault` docs landing pages. See `Testing` section for details.

[task]: https://app.asana.com/0/1202097197789424/1202764970659506/f
[preview]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/
[/boundary/api-docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/boundary/api-docs
[/boundary/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/boundary/docs
[/consul/api-docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/consul/api-docs
[/consul/commands]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/consul/commands
[/consul/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/consul/docs
[/hcp/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/hcp/docs
[/nomad/api-docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/nomad/api-docs
[/nomad/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/nomad/docs
[/nomad/intro]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/nomad/intro
[/nomad/plugins]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/nomad/plugins
[/nomad/tools]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/nomad/tools
[/packer/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/packer/docs
[/packer/guides]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/packer/guides
[/packer/intro]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/packer/intro
[/packer/plugins]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/packer/plugins
[/terraform/cdktf]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/cdktf
[/terraform/cli]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/cli
[/terraform/cloud-docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/cloud-docs
[/terraform/cloud-docs/agents]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/cloud-docs/agents
[/terraform/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/docs
[/terraform/enterprise]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/enterprise
[/terraform/internals]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/internals
[/terraform/intro]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/intro
[/terraform/language]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/language
[/terraform/plugin]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/plugin
[/terraform/plugin/framework]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/plugin/framework
[/terraform/plugin/log]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/plugin/log
[/terraform/plugin/mux]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/plugin/mux
[/terraform/plugin/sdkv2]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/plugin/sdkv2
[/terraform/registry]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/terraform/registry
[/vagrant/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/vagrant/docs
[/vagrant/intro]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/vagrant/intro
[/vagrant/vagrant-cloud]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/vagrant/vagrant-cloud
[/vault/api-docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/vault/api-docs
[/vault/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/vault/docs
[/waypoint/commands]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/commands
[/waypoint/docs]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/docs
[/waypoint/plugins]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/plugins

### CODE REVIEW

File: src/components/landing-hero/index.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { useCurrentProduct } from 'contexts'
+import IconTileLogo from 'components/icon-tile-logo'
+import Heading from 'components/heading'
+import Text from 'components/text'
+import { LandingHeroProps } from './types'
+import s from './landing-hero.module.css'
+
+const LandingHero = ({ pageHeading, pageSubtitle }: LandingHeroProps) => {
```
Comment (zchsh) [inline src/components/landing-hero/index.tsx:8]: This, and the associated CSS, hasn't been changed at all relative to #733. Presentation layer FTW 🙌

File: src/data/terraform.json [hunk 2]
```diff
@@ -106,7 +106,8 @@
 			"name": "Configuration Language",
 			"path": "language",
 			"visuallyHideSidebarTitle": true,
-			"addOverviewItem": false
+			"addOverviewItem": false,
+			"description": "This a custom description of the Terraform Language page that is written in src/data/terraform.json. If this weren't here, we'd fall back to the frontmatter description."
```
Comment (zchsh) [inline src/data/terraform.json:110]: Added in this custom description as a demo. Expectation is that we'll sort out this content at the same time we address other Terraform placeholder content.

File: src/types/products.ts [hunk 3]
```diff
@@ -131,6 +131,14 @@ interface RootDocsPath {
...
 	 */
 	addOverviewItem?: boolean
+
+	/**
+	 * An optional description for this category of documentation.
+	 * Shown as the subtitle of the docs landing hero element.
+	 * If omitted, falls back to the page's authored frontMatter.description,
+	 * or falls back to an empty string.
+	 */
+	description?: string
```
Comment (zchsh) [inline src/types/products.ts:141]: Have made this `description` property, and am falling back to `frontMatter.description`, based on [input from Judith in Slack](https://hashicorp.slack.com/archives/C01KCU4HDPY/p1661181737637109?thread_ts=1660952605.279929&cid=C01KCU4HDPY) that authors will likely want to use `frontMatter.description`.

File: src/views/docs-view/components/docs-root-landing/index.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+export default function DocsRootLanding({
+	children,
+	subtitle,
+}: {
+	children: ReactChild[]
+	subtitle: string
+}) {
+	/**
+	 * Iterate over the content children,
+	 * extracting <h1> text to use as the `heading` for our `LandingHero`,
```
Comment (zchsh) [inline src/views/docs-view/components/docs-root-landing/index.tsx:22]: I briefly considered the option of writing a `remark` plugin to remove the `h1` node, so that we don't have to `Children.map` on each render of content. However, felt like a bit of an optimization, and wanted to get any opinions on whether that would feel worthwhile, so opted not to do it for now.

File: src/views/docs-view/docs-view.module.css [hunk 5]
```diff
@@ -0,0 +1,3 @@
+.mdxContent {
+	margin-top: 48px;
```
Comment (zchsh) [inline src/views/docs-view/docs-view.module.css:2]: Long-winded note on spacing in CSS:

> TLDR I'm trying to stick to `margin-top`. And avoid components defining the spacing around themselves, [per Notion](https://www.notion.so/hashicorp/CSS-Modules-753f2647f7aa4a598e28990d9f0d05a8).

---


I had trouble reasoning around spacing, as we had a combination of padding and margin across various `top` and `bottom` properties. Our [Notion guidelines](https://www.notion.so/hashicorp/CSS-Modules-753f2647f7aa4a598e28990d9f0d05a8) mention `No space around components`, so I started with that. I think a decent additional guidelines would be to consistently apply spacing between elements with EITHER `margin/padding-top` or `margin/padding-bottom` - having a mix of the two can make it difficult to reason about spacing, as we have to think about combining pairs of elements. I've gone with `margin-top` on the second element, as I find it tends to be more predictable with respect to optional elements and spacing at the end of a section.

Happy to dive further into spacing issues, this is definitely an area I would love to get involved in in terms of finding consistent practices, which I think would be best to establish across design and engineering so that we have a shared mental model. For now I'm not trying to dive too deep so as to keep the scope of this PR reasonable.

File: src/views/docs-view/components/docs-root-landing/utils/get-element-text-content.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { Children, ReactChild, ReactElement } from 'react'
+
+/**
+ * Given a React element,
+ * Return the text content of that element.
+ */
+export function getElementTextContent(element: ReactElement) {
```
Comment (zchsh) [inline src/views/docs-view/components/docs-root-landing/utils/get-element-text-content.ts:7]: No changes to this utility relative to #733, have just moved it to a separate file.

File: src/views/docs-view/index.tsx [hunk 7]
```diff
@@ -56,14 +58,27 @@ const productsToPrimitives: ProductsToPrimitivesMap = {
...
 	const currentProduct = useCurrentProduct()
 	const { compiledSource, scope } = mdxSource
 	const additionalComponents = productsToPrimitives[currentProduct.slug] || {}
 	const components = defaultMdxComponents({ additionalComponents })
 	const shouldRenderSearch =
 		!hideSearch && __config.flags.enable_product_docs_search
 
+	const Layout =
+		layouts[metadata?.layout?.name] ??
+		(({ children }) => <div className={s.mdxContent}>{children}</div>)
```
Comment (zchsh) [inline src/views/docs-view/index.tsx:80]: No changes here relative to #733, except the use of `s.mdxContent` for spacing. Comment on `docs-view.module.css` has a bit more detail on spacing thoughts.

File: src/views/docs-view/server.ts [hunk 8]
```diff
@@ -345,11 +345,30 @@ export function getStaticGenerationFunctions<
 				layoutProps.githubFileUrl = githubFileUrl
 			}
 
+			/**
+			 * TODO: this logic is temporary to showcase how layouts can work.
+			 * Going forward this would be defined in frontmatter,
+			 * we'd pass frontmatter.layout directly rather than constructing this.
+			 *
+			 * Task: https://app.asana.com/0/1202097197789424/1202850056121889/f
```
Comment (zchsh) [inline src/views/docs-view/server.ts:353]: Grouped the `layout` shim together a bit here, and added an Asana task for further `layout` work follow-up. Other than that, this is effectively identical to what was in #733.

**Comment by zchsh** [comment]
@BRKalow  This PR implements custom docs landing pages.

It's a fairly light refactor of your work in #733. Key difference is:
- The first `<p/>` is no longer extracted for `subheading` of the hero
    - Instead, we use either `currentRootDocsPath.description` (a new optional property) or fall back to `frontMatter.description` (a suggestion [from Judith](https://hashicorp.slack.com/archives/C01KCU4HDPY/p1661181737637109?thread_ts=1660952605.279929&cid=C01KCU4HDPY))
    - In some cases, notably [/waypoint/commands][/waypoint/commands] and [/waypoint/plugins][/waypoint/plugins]

---

@braydenlove  tagging you here as well, since this affects the existing Vault and Waypoint landing pages where the "docs landing hero" was not yet implemented. I couldn't track down design specs for those so would definitely love to get your eyes on this 🙇‍♂️

---

cc @mhonnatti as well, I expect we may need some content co-ordination here to get the hero elements on [/waypoint/commands][/waypoint/commands] and [/waypoint/plugins][/waypoint/plugins] filled in.

---

> Note: latest build failed due to Packer plugins, latest commit was adding a comment line linking to an Asana task, so I'm fairly confident we'll be able to resolve with rebuilds.

[/waypoint/commands]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/commands
[/waypoint/plugins]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/plugins

**Comment by braydenlove** [comment]
@zchsh [here](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Developer?node-id=11028%3A104591) are the heroes! Whatever copy is fine, I just grabbed something to get the conversation started. Happy to do whatever there <3.

**Comment by zchsh** [comment]
@braydenlove Ah awesome, ty! 💖 I think these are matching the spec, but would love your VQA on them if you have the bandwidth.

Potential content gap is probably the main thing here then, you'd mentioned we wouldn't want these hero elements without subheadings, but in the absence of content that's the state we're in on [/waypoint/commands][/waypoint/commands] and [/waypoint/plugins][/waypoint/plugins] pages currently:

CleanShot 2022-08-23 at 17 39 48@2x

CleanShot 2022-08-23 at 17 40 05@2x

Let me know if you think those would be acceptable to ship without a `subheading`, or if we need to pause on this to get content sorted out 🙇 

[/waypoint/commands]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/commands
[/waypoint/plugins]: https://dev-portal-git-brkzscustom-docs-hero-hashicorp.vercel.app/waypoint/plugins

File: src/views/docs-view/server.ts [hunk 9]
```diff
@@ -345,11 +345,30 @@ export function getStaticGenerationFunctions<
...
+							null,
+				  }
+				: null
+
 			const finalProps = {
 				layoutProps,
 				metadata: {
 					title: frontMatter.page_title ?? null,
 					description: frontMatter.description ?? null,
+					layout: customLayout,
```
Comment (BRKalow) [inline src/views/docs-view/server.ts:371]: 💭 I think this file is one of our biggest ones in the project, what do you think about abstracting out this logic?

```ts
layout: getCustomLayout(...)
```

File: src/views/docs-view/index.tsx (hunk 7 above)
Comment (BRKalow) [reply src/views/docs-view/index.tsx:80]: suggestion: let's pull the default layout declaration out of the component body, otherwise re-renders of the view will cause the layout component to re-mount.

```ts
const DefaultLayout = (({ children }) => <div className={s.mdxContent}>{children}</div>
```

File: src/views/docs-view/components/docs-root-landing/index.tsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+		// Return non-<h1> elements unmodified
+		return child
+	})
+
+	/**
+	 * MDX content should contain an <h1/>, but this is not guaranteed.
+	 * Our <LandingHero /> requires a heading, so we avoid rendering it
+	 * if we have not found an <h1> in the MDX content.
+	 */
```
Comment (BRKalow) [inline src/views/docs-view/components/docs-root-landing/index.tsx:49]: 👍

File: src/views/docs-view/index.tsx [hunk 11]
```diff
@@ -1,3 +1,4 @@
+import { Fragment } from 'react'
```
Comment (BRKalow) [inline src/views/docs-view/index.tsx:1]: nit 👀 : this looks unused now

File: src/views/docs-view/docs-view.module.css (hunk 5 above)
Comment (BRKalow) [reply src/views/docs-view/docs-view.module.css:2]: I generally agree re spacing considerations, but have preferred `margin-bottom` for no other reason than preference 😄 Thanks for the insight here!

**Review by BRKalow — COMMENTED** [review]
This looks great, really nicely done! Thanks for taking this one over the finish line🙇

**Comment by braydenlove** [comment]
@zchsh thanks for the visual! Yeah it looks funky but is what it is for now? Curious if it would look better to center it but... Let me spin on it tonight and I will get back to you tomorrow!

File: src/views/docs-view/server.ts (hunk 9 above)
Comment (zchsh) [reply src/views/docs-view/server.ts]: 💯 nice, I dig it - ✅ updated!

File: src/views/docs-view/index.tsx (hunk 7 above)
Comment (zchsh) [reply src/views/docs-view/index.tsx]: 🙌  ✅ thanks for catching this, updated!

File: src/views/docs-view/index.tsx (hunk 11 above)
Comment (zchsh) [reply src/views/docs-view/index.tsx]: 🔪 ✅

**Comment by zchsh** [comment]
@BRKalow @braydenlove  thanks for the feedback here 💖 I've update the PR as follows:

## Implementation details

- pulled `DefaultLayout` out of component body to prevent re-mounting on each render
- splits out `getCustomLayout` from `docs-view/server.ts`, to avoid that file getting too big
- removed unused import in `docs-view`

## Design tweaks

- added `align-items: center` styling for when no `pageSubtitle` is present
    - if `pageSubtitle` is present, no changes to styling

I might've jumped the gun here a bit @braydenlove , but this was some quick CSS to write so I figured I'd put something out there in case it saves some Figma mockup effort. Definitely happy to adjust if you have more specific specs you'd want to see here 🙇‍♂️

Here's a preview of the revised styling when `pageSubtitle` isn't present:

CleanShot 2022-08-24 at 09 45 34@2x

CleanShot 2022-08-24 at 09 45 47@2x

And here's a quick confirmation screenshot that styles are unchanged when `pageSubtitle` is present:

CleanShot 2022-08-24 at 09 45 56@2x

File: src/views/docs-view/utils/get-custom-layout.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+	const isLanding = pathParts.length === 0
+	const customLayout = isLanding
+		? {
+				name: 'docs-root-landing',
+				subtitle:
+					currentRootDocsPath.description ?? frontMatter.description ?? null,
+		  }
+		: null
+	return customLayout
+}
```
Comment (BRKalow) [inline src/views/docs-view/utils/get-custom-layout.ts:28]: 💭 optional suggestion: if we wanna merge this without production impact, we could put it behind a flag for now at this point! We can refine and iterate on the empty state as a follow-up this way.

```suggestion
export function getCustomLayout({
	currentRootDocsPath,
	frontMatter,
	pathParts,
}: {
	currentRootDocsPath: RootDocsPath
	frontMatter: Record<string, unknown>
	pathParts: string[]
}) {
  if (!__config.flags.enable_custom_layouts) {
    return null
  }
	const isLanding = pathParts.length === 0
	const customLayout = isLanding
		? {
				name: 'docs-root-landing',
				subtitle:
					currentRootDocsPath.description ?? frontMatter.description ?? null,
		  }
		: null
	return customLayout
}
```

**Review by BRKalow — APPROVED** [review]
LGTM from an implementation perspective! Left a comment about using a feature flag if we would prefer to refine the empty state as a follow-up.

**Comment by braydenlove** [comment]
@zchsh not jumping the gun at all, awesome work. Thank you!!!

**Review by braydenlove — APPROVED** [review]
!!!

File: src/views/docs-view/utils/get-custom-layout.ts (hunk 12 above)
Comment (zchsh) [reply src/views/docs-view/utils/get-custom-layout.ts:28]: Sweet, I really dig this 💖  In this specific case, I think it feels best to merge in with the empty state we have so far - intent being to avoid getting into a state where we have to kind of "clean up" feature flags later (that would feel a bit like papercut-ish debt!). This is definitely based on the recent design approval below though too - thanks @braydenlove !

I'll definitely keep feature flags in mind for future work 👍 💯
